[JENKINS-34681] Allow custom plugin managers when using war

andresrc opened the pull request:
See [JENKINS-34681](https://issues.jenkins-ci.org/browse/JENKINS-34681).

@reviewbybees

andresrc commented on core/src/main/java/hudson/LocalPluginManager.java:
```diff
@@ -37,12 +38,16 @@
+    public LocalPluginManager(ServletContext context, File rootDir) {
+        super(context, new File(rootDir,"plugins"));
+    }
+
```
New constructor to simplify extension of this class.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -262,6 +264,25 @@ public void joinInit() throws InterruptedException {
+     * Creates the plugin manager to use, if customized.
+     * @return The plugin manager to use or {@code null} to use the default one.
+     */
+    private PluginManager createPluginManager(ServletContext context, File rootDir) {
```
Would be nice to add CheckForNull and to annotate Hudson/Jenkins constructors. NIT for this PR

andresrc replied:
Added the annotations here, additional ones left for a downstream PR.

oleg-nenashev commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -262,6 +264,25 @@ public void joinInit() throws InterruptedException {
+     */
+    private PluginManager createPluginManager(ServletContext context, File rootDir) {
+        final String pmClassName = context.getInitParameter(PluginManager.class.getName() + ".className");
+        if (pmClassName == null) {
```
StringUtils.isEmpty() ?

oleg-nenashev commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -262,6 +264,25 @@ public void joinInit() throws InterruptedException {
+            return null;
+        }
+        try {
+            final Class<? extends PluginManager> klass = Class.forName(pmClassName).asSubclass(PluginManager.class);
```
Class.forName(pmClassName) may return NULL => there may be NPE. It will be handled, but NPEs are always a bad practice

andresrc replied:
Umm, I had never seen `forName` returning null, but many custom class loaders can be involved in this kind of context. Ok.

andresrc replied:
But in any case, whoever is returning `null` would be breaking the class contract...

oleg-nenashev commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -262,6 +264,25 @@ public void joinInit() throws InterruptedException {
+        try {
+            final Class<? extends PluginManager> klass = Class.forName(pmClassName).asSubclass(PluginManager.class);
+            final Constructor <? extends PluginManager> constructor = klass.getConstructor(ServletContext.class, File.class);
+            return constructor.newInstance(context, rootDir);
```
Again, null check with proper logging would be useful

andresrc replied:
According to the [doc](http://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getConstructor(java.lang.Class...)), in case the method does not exist `NoSuchMethodException` is thrown, no null is returned.

oleg-nenashev replied:
Agreed. NIT in such case

oleg-nenashev commented:
Maybe we could test the method by routing JenkinsRule through it...

oleg-nenashev commented on core/src/main/java/hudson/LocalPluginManager.java:
```diff
@@ -37,12 +38,16 @@
  * @author Kohsuke Kawaguchi
  */
 public class LocalPluginManager extends PluginManager {
+    public LocalPluginManager(ServletContext context, File rootDir) {
```
NIT: Javadoc?

batmat commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -262,6 +264,25 @@ public void joinInit() throws InterruptedException {
+            final Class<? extends PluginManager> klass = Class.forName(pmClassName).asSubclass(PluginManager.class);
+            final Constructor <? extends PluginManager> constructor = klass.getConstructor(ServletContext.class, File.class);
+            return constructor.newInstance(context, rootDir);
+        } catch(Exception e) {
```
Wouldn't it be better to catch only `ReflectiveOperationException` here?

As @oleg-nenashev says it's currently also catching NPEs, which isn't actually a good thing IMO too.

amuniz replied:
IMO catch & log is ok in this context. Whatever it happens here we want to continue by falling back to default (built-in plugin manager).

oleg-nenashev replied:
IMHO it's not OK, because a user will be getting crappy NPE warnings in the system log. And it will be likely considered as a bug

oleg-nenashev commented:
🐛 for NPE handling, needs a quick fix

oleg-nenashev commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -267,19 +268,27 @@ public void joinInit() throws InterruptedException {
+        } catch(NullPointerException e) {
+            // Class.forName and Class.getConstructor are supposed to never return null, but just in case we introduce
+            // this specific catch to avoid polluting the logs with NPEs.
+            LOGGER.log(WARNING, String.format("Unable to instantiate custom plugin manager %s. Using default.", pmClassName));
```
Oh well, it was my mistake. Let this code to be a shame for me for years :)

oleg-nenashev commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -267,19 +268,27 @@ public void joinInit() throws InterruptedException {
+        } catch(ClassCastException e) {
+            LOGGER.log(WARNING, String.format("Provided class %s does not extend PluginManager. Using default.", pmClassName));
+        } catch(NoSuchMethodException e) {
+            LOGGER.log(WARNING, String.format("Provided custom plugin manager %s does not provided (ServletContext, File) constructor. Using default.", pmClassName));
```
NIT: I would still expose the original stacktrace. Just in case there is any hidden binary issue for custom cores

oleg-nenashev commented:
🐝 with minor comments

andresrc commented:
Improved test (running it through `JenkinsRule`), addressed comments and squashed.

@reviewbybees again esp. @oleg-nenashev

andresrc commented on test/src/test/java/hudson/CustomPluginManagerTest.java:
```diff
@@ -0,0 +1,157 @@
+ */
+public class CustomPluginManagerTest {
+
+    // TODO: to be greatly simplified when JENKINS-34701 is available.
```
See https://github.com/jenkinsci/jenkins-test-harness/pull/19

andresrc commented on test/src/test/java/hudson/CustomPluginManagerTest.java:
```diff
@@ -0,0 +1,157 @@
+     *
+     * @author Kohsuke Kawaguchi
+     */
+    private static final class NoListenerConfiguration extends AbstractLifeCycle {
```
Also to be removed with https://github.com/jenkinsci/jenkins-test-harness/pull/19

oleg-nenashev commented:
re-:bee:

oleg-nenashev commented:
CC @jenkinsci/code-reviewers

oleg-nenashev commented on core/src/main/java/hudson/WebAppMain.java:
```diff
@@ -262,6 +265,33 @@ public void joinInit() throws InterruptedException {
+     * @return The plugin manager to use or {@code null} to use the default one.
+     */
+    private @CheckForNull PluginManager createPluginManager(ServletContext context, File rootDir) {
+        final String pmClassName = context.getInitParameter(PluginManager.class.getName() + ".className");
```
Actually, why not just System.getProperty() ? Is there a specific purpose of such construct?

oleg-nenashev commented:
@andresrc The current expression does not work for command-line system, properties at least. Switch to <code>System.getProperty()</code> works well
Seems to be a 🐛

andresrc commented:
@oleg-nenashev it was intentional, initially I wanted only to be configurable inside the war, but I can change it to support both mechanisms if it is better, no problem.

oleg-nenashev commented:
@andresrc Yes, it would be better. Maybe we need a kind of such two-way approach for all UpdateManager and UpdateSite properties.

Another problem is a conflict of interfaces. <code>rootDir</code> means different things for PluginManager and classes being initialized automatically. In first case it's JENKINS_HOME/plugins, in second - JENKINS_HOME. Looks like a trap for API users

oleg-nenashev commented:
Should I say that I've successfully fallen into this trap? :)

andresrc commented:
@oleg-nenashev we can escape that trap requiring `LocalPluginManager` as the required parent, in which `rootDir` is `JENKINS_HOME`and translates up the hierarchy.

andresrc commented:
@oleg-nenashev I have identified a couple changes more, let me a few hours.

daniel-beck commented:
:-1: We made `PluginManager` `StaplerOverridable` last year in #1788, and it was never even used. Now there's another core change related to this? I want to see a real justification here and not just a placeholder issue.

oleg-nenashev commented:
Related fix #2332

andresrc commented:
@oleg-nenashev reworked the PR to follow the same approach as https://github.com/jenkinsci/jenkins/pull/2332 (beware, the property name has changed).

@daniel-beck I explained the intended use in [JENKINS-34681](https://issues.jenkins-ci.org/browse/JENKINS-34681)

oleg-nenashev commented:
@andresrc Please do not squash till the review finishes. It's impossible to review diffs in such way

andresrc commented:
@oleg-nenashev sorry, this time the changes were too big and thought it was better to do it now. We'll leave it for the end the review next time.

oleg-nenashev commented on core/src/main/java/hudson/LocalPluginManager.java:
```diff
@@ -24,25 +24,93 @@
-    public LocalPluginManager(Jenkins jenkins) {
-        super(jenkins.servletContext, new File(jenkins.getRootDir(),"plugins"));
+    /** Custom plugin manager system property or context param. */
+    public static final String CUSTOM_PLUGIN_MANAGER = PluginManager.class.getName() + ".className";
```
Actually it still points to <code>PluginManager</code>
Works well for me

oleg-nenashev commented on core/src/main/java/hudson/LocalPluginManager.java:
```diff
@@ -24,25 +24,93 @@
  *
  * @author Kohsuke Kawaguchi
  */
 public class LocalPluginManager extends PluginManager {
```
I still see no real need in this class, but oh well

oleg-nenashev commented on core/src/main/java/hudson/LocalPluginManager.java:
```diff
@@ -24,25 +24,93 @@
+     * @param context Servlet context. Provided for compatibility as {@code Jenkins.getInstance().servletContext} should be used.
+     * @param rootDir Jenkins home directory.
+     */
+    public LocalPluginManager(ServletContext context, @NonNull File rootDir) {
```
CheckForNull ?

oleg-nenashev commented:
Looks good to me. 🐝 with minor comments

andresrc commented:
@oleg-nenashev I'm still giving a thought on the base class and constructors to require...

andresrc commented:
@oleg-nenashev final form now, will squash after the review.

oleg-nenashev commented:
@andresrc 
> will squash after the review. 

No real nead. It can be easily done by a person, who performs merge

oleg-nenashev commented:
re-:bee: if the test passes

varmenise commented:
🐝

andresrc commented:
@reviewbybees done

ping @jenkinsci/code-reviewers

batmat commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -144,10 +147,69 @@
+ * (e.g. Jenkins module or Extra library dependency in the WAR file project).
+ * Plugins cannot be used for such purpose.
+ * In order to be correctly instantiated, the class definition must have at least one constructor with the same
+ * siganture as the following ones:
```
typo: signature

batmat commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -144,10 +147,69 @@
+
+    /**
+     * Creates the {@link PluginManager} to use if no one is provided to a {@link Jenkins} object.
+     * This method will be after creation of {@link Jenkins} object, but before it is fully initialized.
```
will be /called/ ? missing word?

batmat commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -144,10 +147,69 @@
+     * This method will be after creation of {@link Jenkins} object, but before it is fully initialized.
+     * @param jenkins Jenkins Instance.
+     * @return Plugin manager to use. If no custom class is configured or in case of any error, the default
+     * {@link LocalPluginManager} returned.
```
/is/ returned?

batmat commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -144,10 +147,69 @@
+                        return klass.getConstructor(ServletContext.class, File.class).newInstance(jenkins.servletContext, jenkins.getRootDir());
+                    } catch(NoSuchMethodException e2) {
+                        return klass.getConstructor(File.class).newInstance(jenkins.getRootDir());
+                    }
```
That nesting level makes me quite uneasy.

oleg-nenashev commented:
@batmat would it be OK if I address the comments in a separate PR?

andresrc commented:
@oleg-nenashev @batmat I can revisit the typos an provide better handling of the the constructors in a separate PR that also integrates https://github.com/jenkinsci/jenkins/pull/2337 as soon as it is merged.

Would that be ok?

oleg-nenashev commented:
+1 from me

batmat commented:
@oleg-nenashev @andresrc sure, please do. 

If I had seen something important, I would have put a :-1: my comments above were mostly NIT. Maybe only the "too deeply nested" part is more than NIT.

oleg-nenashev commented:
Now there is a merge conflict

oleg-nenashev commented:
@andresrc When you integrate the changes, could you please migrate all properties in PluginManager to SystemProperties? It would allow to avoid merge conflicts

Commit:
[JENKINS-34681] Allow custom default plugin managers

andresrc commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -561,7 +654,7 @@ protected static void addDependencies(URL hpiResUrl, String fromPath, Set<URL> d
                     addDependencies(dependencyURL, fromPath, dependencySet);
-                    // And then add the current plugin 
+                    // And then add the current plugin
                     dependencySet.add(dependencyURL);
```
I know, unrelated, and have to review the IDE config :(

andresrc commented:
* [x] Addressed typos.
* [x] Review structure when trying different constructors (cc @batmat )
* [x] More tests (all constructors and invalid constructor covered)
* [x] Rebased

Commit:
[JENKINS-34681] Convert remaining properties to SystemProperties.

oleg-nenashev commented:
re-:bee:

oleg-nenashev commented:
merging since there were only small changes since the previous review cycle
